### ISSUE TITLE
Updating array inside FormGroupState in reducer

### ISSUE THREAD

**Issue by philjones88** [MrWolfZ/ngrx-forms#17]
As per title assume I have this reducer:

```
export class TemplateItemViewModel {
    public foo: string;
    public bar: string:
}

export class TemplateViewModel {
    public name: string;
    public description: string;
    public items: TemplateItemViewModel[];
}

export const initialState = createFormGroupState<TemplateViewModel>(
    'template',
    new TemplateViewModel()
);

export function reducer(state = initialState, action: editTemplateActions.Actions): FormGroupState<TemplateViewModel> {
    const myForm = formGroupReducer(state, action);
    if (myForm !== state) {
        state = { ...state, ...myForm };
    }

    switch (action.type) {

        case editTemplateActions.FETCH_COMPLETE:
            return createFormGroupState<TemplateViewModel>(
                'template',
                action.payload
            );

        case editTemplateActions.ADD_ROLE:
            // ???
            return state;

        default:
            return state;
    }
}
```

The problem becomes the documentation (https://github.com/MrWolfZ/ngrx-forms/blob/master/docs/FORM_ARRAYS.md) talks about updating a `FormArrayState<T>` not an array part of `FormGroupState<T>`.

I've tried using `setValue` but it triggers ngrx freeze to throw an error.

Do you have any examples? I've read the documentation and the example app, perhaps I'm missing something obvious!

Thanks, Phil

**Comment by philjones88** [comment]
Not sure this is the best approach but I got it working with:

```
export function createState(model: TemplateViewModel) {
    return createFormGroupState<TemplateViewModel>(
        'template',
        model
    );
}

        case editTemplateActions.ADD_ROLE:
            const newState: TemplateViewModel = { ...state.value, items: [ ...state.value.items, new TemplateRoleViewModel() ] };

            return createState(newState);
```

I'm guessing it's remaking the entire state tree but it works for now and I can change it later.

**Comment by MrWolfZ** [comment]
So you want to add a new control for a `TemplateItemViewModel` to the array if the `ADD_ROLE` action is dispatched? That would look like this (I haven't actually run the code so it may contain compile errors):

```typescript
// there is really no need to use classes; instead just use interfaces and
// plain objects since the class identity is not guaranteed to be preserved
// inside ngrx-forms (basically the first time the value changes your class
// information is lost since it's properties are copied into a plain object)
export interface TemplateItemViewModel {
  foo: string;
  bar: string:
}

export interface TemplateViewModel {
  name: string;
  description: string;
  items: TemplateItemViewModel[];
}

// initialize your state with all the properties you want in the form, i.e. make
// sure it is fully initialized to prevent issues in your template because controls
// are missing even though your interface states they are always present
export const initialState = createFormGroupState<TemplateViewModel>(
  'template',
  {
    name: '',
    description: '',
    items: [],
  },
);

export function reducer(state = initialState, action: editTemplateActions.Actions) {
  // since this reducer is only reducing the form state and nothing
  // else, there's no need for the inequality check from your example,
  // since that is all handled internally by ngrx-forms already
  state = formGroupReducer(state, action);

  switch (action.type) {

    case editTemplateActions.FETCH_COMPLETE:
      return createFormGroupState<TemplateViewModel>(
        'template',
        action.payload,
      );

    case editTemplateActions.ADD_ROLE:
      // to update a form array inside a group just nest the update function
      // you want to run on the array inside an updateGroup call
      return updateGroup<TemplateViewModel>(state, {
        items: addArrayControl(''),
      });

    default:
      return state;
  }
}
```

**Comment by philjones88** [comment]
@MrWolfZ thanks for the example, however I get an exception when trying to implement it in my application. I suspects ngrx freeze was to blame but it's not.

```
isArrayState@webpack-internal:///../../../../ngrx-forms/ngrx/forms.es5.js:352:5
addArrayControl@webpack-internal:///../../../../ngrx-forms/ngrx/forms.es5.js:2423:9
reducer@webpack-internal:///../../../../../src/app/system/cospace-template/reducers/edit.ts:29:108
combination@webpack-internal:///../../../../@ngrx/store/@ngrx/store.es5.js:96:52
combination@webpack-internal:///../../../../@ngrx/store/@ngrx/store.es5.js:96:52
freeze@webpack-internal:///../../../../ngrx-store-freeze/src/index.js:12:25
exports.localStorageSync/</<@webpack-internal:///../../../../ngrx-store-localstorage/dist/index.js:167:25
computeNextEntry@webpack-internal:///../../../../@ngrx/store-devtools/@ngrx/store-devtools.es5.js:296:21
recomputeStates@webpack-internal:///../../../../@ngrx/store-devtools/@ngrx/store-devtools.es5.js:335:15
liftReducerWith/</<@webpack-internal:///../../../../@ngrx/store-devtools/@ngrx/store-devtools.es5.js:530:26
StoreDevtools/liftedStateSubscription<@webpack-internal:///../../../../@ngrx/store-devtools/@ngrx/store-devtools.es5.js:595:33
ScanSubscriber.prototype._tryNext@webpack-internal:///../../../../rxjs/_esm5/operators/scan.js:116:22
ScanSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/operators/scan.js:109:20
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
WithLatestFromSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/operators/withLatestFrom.js:119:17
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
Notification.prototype.observe@webpack-internal:///../../../../rxjs/_esm5/Notification.js:34:41
ObserveOnSubscriber.dispatch@webpack-internal:///../../../../rxjs/_esm5/operators/observeOn.js:101:9
AsyncAction.prototype._execute@webpack-internal:///../../../../rxjs/_esm5/scheduler/AsyncAction.js:122:13
QueueAction.prototype.execute@webpack-internal:///../../../../rxjs/_esm5/scheduler/QueueAction.js:39:13
AsyncScheduler.prototype.flush@webpack-internal:///../../../../rxjs/_esm5/scheduler/AsyncScheduler.js:40:25
QueueAction.prototype.schedule@webpack-internal:///../../../../rxjs/_esm5/scheduler/QueueAction.js:33:9
Scheduler.prototype.schedule@webpack-internal:///../../../../rxjs/_esm5/Scheduler.js:47:16
ObserveOnSubscriber.prototype.scheduleMessage@webpack-internal:///../../../../rxjs/_esm5/operators/observeOn.js:105:18
ObserveOnSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/operators/observeOn.js:108:9
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
MergeMapSubscriber.prototype.notifyNext@webpack-internal:///../../../../rxjs/_esm5/operators/mergeMap.js:156:13
InnerSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/InnerSubscriber.js:27:9
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
MapSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/operators/map.js:88:9
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
MergeMapSubscriber.prototype.notifyNext@webpack-internal:///../../../../rxjs/_esm5/operators/mergeMap.js:156:13
InnerSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/InnerSubscriber.js:27:9
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
SkipSubscriber.prototype._next@webpack-internal:///../../../../rxjs/_esm5/operators/skip.js:49:13
Subscriber.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subscriber.js:96:13
Subject.prototype.next@webpack-internal:///../../../../rxjs/_esm5/Subject.js:65:17
BehaviorSubject.prototype.next@webpack-internal:///../../../../rxjs/_esm5/BehaviorSubject.js:49:9
ActionsSubject.prototype.next@webpack-internal:///../../../../@ngrx/store/@ngrx/store.es5.js:176:9
Store.prototype.dispatch@webpack-internal:///../../../../@ngrx/store/@ngrx/store.es5.js:452:9
EditComponent.prototype.addRole@webpack-internal:///../../../../../src/app/system/cospace-template/components/edit/edit.component.ts:71:9
View_EditComponent_3/<@ng:///CospaceTemplateModule/EditComponent.ngfactory.js:74:23
handleEvent@webpack-internal:///../../../core/esm5/core.js:13451:132
callWithDebugContext@webpack-internal:///../../../core/esm5/core.js:14936:39
debugHandleEvent@webpack-internal:///../../../core/esm5/core.js:14523:12
dispatchEvent@webpack-internal:///../../../core/esm5/core.js:9900:16
renderEventHandlerClosure/<@webpack-internal:///../../../core/esm5/core.js:10514:38
decoratePreventDefault/<@webpack-internal:///../../../platform-browser/esm5/platform-browser.js:2665:53
ZoneDelegate.prototype.invokeTask@webpack-internal:///../../../../zone.js/dist/zone.js:425:17
onInvokeTask@webpack-internal:///../../../core/esm5/core.js:4816:24
ZoneDelegate.prototype.invokeTask@webpack-internal:///../../../../zone.js/dist/zone.js:424:17
Zone.prototype.runTask@webpack-internal:///../../../../zone.js/dist/zone.js:192:28
ZoneTask.invokeTask@webpack-internal:///../../../../zone.js/dist/zone.js:499:24
invokeTask@webpack-internal:///../../../../zone.js/dist/zone.js:1540:9
globalZoneAwareCallback@webpack-internal:///../../../../zone.js/dist/zone.js:1566:17
```

**Comment by MrWolfZ** [comment]
The error seems to only contain the stack trace, not the actual error message. Could you please post the full error?

However, I noticed an error in my example. It should be addArrayControl({ foo: ‘’, bar: ‘’ }) instead of addArrayControl(‘’). Maybe that already solves the issue.

**Comment by philjones88** [comment]
That's not the issue and it's the best trace the Angular CLI + Firefox Quantum devtools give me.

Added some poor mans debugging `console.log`'s to the `addArrayControl` method:

```
function addArrayControl(value, indexOrState, index) {
    console.log('addArrayControl', value, indexOrState, index);
...
}
```

And it seems to be passing the new object in `value` fine but then `undefined` for the `indexOrState` and `index`.

**Comment by MrWolfZ** [comment]
Yup, that’s a bug indeed. The variable being undefined is actually fine, but there’s a missing “null check”. I looked at my tests and I did indeed forget to test this case (shame on me).

I'll push a fix this evening. Until then you can simply replace the addArrayControl call with

s => addArrayControl({ foo: ‘’, bar: ‘’ }, s, s.value.length)

This manually specifies the index to insert the element into.

Thanks for finding this bug.

**Comment by philjones88** [comment]
Thanks, the workaround works.

Slowly wrapping my head around ngrx-forms. I'm not the best Typescript/Redux dev though :)

Side note: chrome shows a better stack trace (saying `isArrayState` has an issue with `hasOwnProperty` on undefined) than the Firefox Quantum dev tools which is disappointing :(

**Comment by MrWolfZ** [comment]
I just released version 2.0.1 that fixes this issue.

**Comment by ng** [comment]
Unsubscribe.
